### DISCUSSION TITLE
[RFC] Build ELKS in OpenBSD

### DISCUSSION THREAD

**Discussion by floriangit** [jbruchon/elks#1894] [Show and tell]
Since I switched to OpenBSD (7.5) on my main PC, I thought to try to build ELKS there...and somehow got it going! My image folder is populated :1st_place_medal: 
There were several quirks obviously, but it might be of interest to get OpenBSD and maybe also other BSDs supported either by a README/WIKI or even better by having this supported in the build itself?
Here are the manual things I had to do:

- You need to install (at least): `flex`, `bison`, `bash`, `gmake`
- `#/bin/bash` is sometimes hardcoded in ELKS, to avoid any hassle I created a symlink from /usr/local/bin/bash where bash lives in the OpenBSD world
- Made an `alias make='gmake'`
- git cloned `xxd` from github, built it (just one source file) and add the executable somewhere in my $`PATH`
- Deleted the loop condition  "`until ... done`" in tools/Makefile since it was looping endlessly wgetting stuff. Why? Not found out yet (see below)
- Actually be in a bash before attempting to `make`
- Used `make config`, since `make menuconfig` needs ncurses

After `tools/build.sh` I had to manually do this (reason? unknown):
```
cd elks/cross/libexec/gcc/ia16-elf/6.3.0/
ln -s liblto_plugin.so.0.0 liblto_plugin.so  
```

Diff:

```
-- a/tools/Makefile
+++ b/tools/Makefile
@@ -33,10 +33,8 @@ $(DISTDIR)/$(BINUTILS_DIST).tar.gz:
               $(DISTDIR)/binutils-ia16-*.zip
        set -e; \
        cd $(DISTDIR); \
-       until gunzip -t $(BINUTILS_DIST).tar.gz.tmp 2>/dev/null; do \
                wget -c https://github.com/tkchia/binutils-ia16/archive/$(BINUTILS_VER).tar.gz \
                     -O $(BINUTILS_DIST).tar.gz.tmp; \
-       done
        cd $(DISTDIR) && mv $(BINUTILS_DIST).tar.gz.tmp $(BINUTILS_DIST).tar.gz
 
 $(BUILDDIR)/.binutils.src: $(DISTDIR)/$(BINUTILS_DIST).tar.gz
@@ -109,10 +107,8 @@ $(DISTDIR)/$(GCC_DIST).tar.gz:
               $(DISTDIR)/gcc-ia16-*.zip
        set -e; \
        cd $(DISTDIR); \
-       until gunzip -t $(GCC_DIST).tar.gz.tmp 2>/dev/null; do \
                wget -c https://github.com/tkchia/gcc-ia16/archive/$(GCC_VER).tar.gz \
                     -O $(GCC_DIST).tar.gz.tmp; \
-       done
        cd $(DISTDIR) && mv $(GCC_DIST).tar.gz.tmp $(GCC_DIST).tar.gz
 
 $(BUILDDIR)/.gcc.src: $(DISTDIR)/$(GCC_DIST).tar.gz \
@@ -166,10 +162,8 @@ $(DISTDIR)/$(EMU86_DIST).tar.gz:
               $(DISTDIR)/emu86-*.zip
        set -e; \
        cd $(DISTDIR); \
-       until gunzip -t $(EMU86_DIST).tar.gz.tmp 2>/dev/null; do \
                wget -c https://github.com/mfld-fr/emu86/archive/$(EMU86_VER).tar.gz \
                     -O $(EMU86_DIST).tar.gz.tmp; \
-       done
        cd $(DISTDIR) && mv $(EMU86_DIST).tar.gz.tmp $(EMU86_DIST).tar.gz

```

**Comment by ghaerr** [comment]
Thanks for the figuring out how to get ELKS built on OpenBSD. We can leave this discussion open so that others can use it.

> git cloned xxd from github, built it (just one source file) and add the executable somewhere in my $PATH

~~I'm noticing that `xxd` is only used in bootblocks/Makefile, and creating a file that isn't even used. I'll remove that for one less dependency.~~

> Deleted the loop condition "until ... done" in tools/Makefile since it was looping endlessly wgetting stuff. 

Not sure on that, perhaps OpenBSD's `gunzip` exits with a non-zero value even when successful, or doesn't support -t?

> Used make config, since make menuconfig needs ncurses

Is ncurses not available as a package-installed library on OpenBSD?

**Reply by floriangit** [reply]
> Not sure on that, perhaps OpenBSD's gunzip exits with a non-zero value even when successful, or doesn't support -t?

I've checked a bit on this. `-t` is supported on linux and obsd and both descriptions ("only test integrity (Linux: and exit)") read similar. Then I tested to manually check a) until behavior (since different bash versions) and b) gunzip exit codes. The only diff I found is, that gunzip's exit code when unsuccessful is 2 in obsd and 1 in linux...it's 0 on success everywhere, so for me this is still a bit mysterious...

> Is ncurses not available as a package-installed library on OpenBSD?

Not installable in the official installation, no. There might be ports out there, but I did not try hard, since the amount of ELKS kernel items is still manageable with a `make config`.

**Reply by ghaerr** [reply]
> since the amount of ELKS kernel items is still manageable with a make config.

I usually just edit .config itself for minor changes. We've finally whittled the configuration file down and its even manageable manually now and can give a good overview of what is set/unset.

**Comment by Oichkatzelesfrettschen** [comment]
Silly question, but it seems you are quite the BSD person. Do you know if this can be built on NetBSD as well, or would instructions be modified? 

My understanding (which is usually 15% correct) is that NetBSD is "lighter weight" for low-end systems.

**Reply by floriangit** [reply]
Oh, I'm not a BSD expert at all. And I have not tried on Free- or Net- BSDs. I would not anticipate too much difference, though. Maybe others have tried...?

If ELKS would have wanted to have added BSD as a fully supported build/host platform, I might have tried to work more on that, but Greg was good enough with this mini-FAQ on github. So far.

**Reply by tyama501** [reply]
I am building ELKS with (old) FreeBSD.
I think you can build it with a little modification.
